qiskit_experiments/database_service/db_experiment_data.py

```diff
@@ -0,0 +1,1042 @@
+
+        del self._figures[figure_key]
+        self._deleted_figures.append(figure_key)
+
+        if self._service and self.auto_save:
+            with service_exception_to_warning():
+                self.service.delete_figure(experiment_id=self.experiment_id, figure_name=figure_key)
+            self._deleted_figures.remove(figure_key)
+
+        return figure_key
+
+    def figure(
+        self, figure_key: Union[str, int], file_name: Optional[str] = None
+    ) -> Union[int, bytes]:
+        """Retrieve the specified experiment figure.
+
+        Args:
+            figure_key: Name or index of the figure.
+            file_name: Name of the local file to save the figure to. If ``None``,
+                the content of the figure is returned instead.
+
+        Returns:
+            The size of the figure if `file_name` is specified. Otherwise the
+            content of the figure in bytes.
+
+        Raises:
+            DbExperimentEntryNotFound: If the figure cannot be found.
+        """
+        if isinstance(figure_key, int):
+            figure_key = self._figures.keys()[figure_key]
+
+        figure_data = self._figures.get(figure_key, None)
+        if figure_data is None and self.service:
+            figure_data = self.service.figure(
+                experiment_id=self.experiment_id, figure_name=figure_key
+            )
+            self._figures[figure_key] = figure_data
+
+        if figure_data is None:
+            raise DbExperimentEntryNotFound(f"Figure {figure_key} not found.")
+
+        if file_name:
+            with open(file_name, "wb") as output:
+                num_bytes = output.write(figure_data)
+                return num_bytes
+        return figure_data
+
+    @auto_save
+    def add_analysis_results(
+        self,
+        results: Union[DbAnalysisResult, List[DbAnalysisResult]],
+    ) -> None:
+        """Save the analysis result.
+
+        Args:
+            results: Analysis results to be saved.
+        """
+        if not isinstance(results, list):
+            results = [results]
+
+        for result in results:
+            self._analysis_results[result.result_id] = result
+
+            with contextlib.suppress(DbExperimentDataError):
+                result.service = self.service
+                result.auto_save = self.auto_save
+
+            if self.auto_save and self._service:
+                result.save()
+
+    @auto_save
+    def delete_analysis_result(
+        self,
+        result_key: Union[int, str],
+    ) -> str:
+        """Delete the analysis result.
+
+        Args:
+            result_key: ID or index of the analysis result to be delete.
+
+        Returns:
+            Analysis result ID.
+
+        Raises:
+            DbExperimentEntryNotFound: If analysis result not found.
+        """
+        if isinstance(result_key, int):
+            result_key = self._analysis_results.keys()[result_key]
+        elif result_key not in self._analysis_results:
```

Comment:
Is this checking a cache? If so couldn't this get out of sync with the source of truth (the actual backing service?).

Reply:
Fixed in 2e01e94